### ISSUE TITLE
PyTorch API gaps for browser-first training curriculum

### ISSUE THREAD

**Issue by vijayksingh** [unlocalhosted/browsergrad#5]
## Context

I am using `@unlocalhosted/browsergrad-jit@0.8.1` through Pyodide to run browser-first PyTorch-style educational exercises.

Current coverage is already enough for tensors, `nn.Module`, `nn.Parameter`, `nn.Linear`, `optim.SGD`, `optim.Adam`, MSE loss, basic autograd, and small training-step exercises.

While building a PyTorch study path, a few missing APIs blocked otherwise natural beginner/intermediate units around validation mode, checkpointing, detach/logging boundaries, binary classifiers, and debugging reports.

## Probe

```js
import { loadPyodide } from "pyodide";
import { installJit } from "@unlocalhosted/browsergrad-jit";
import { createNodePyodideTarget } from "@unlocalhosted/browsergrad-jit/node-adapter";

const py = await loadPyodide({ stdout: () => {}, stderr: () => {} });
await py.loadPackage(["numpy"]);
await installJit(createNodePyodideTarget(py));
await py.runPythonAsync("import browsergrad_jit as bg\nbg.install_torch_alias()");

const result = await py.runPythonAsync(`
import torch, json
import torch.nn as nn
caps = {
  'version_context': '@unlocalhosted/browsergrad-jit 0.8.1 via Pyodide',
  'nn.Dropout': hasattr(nn, 'Dropout'),
  'nn.BatchNorm1d': hasattr(nn, 'BatchNorm1d'),
  'torch.no_grad': hasattr(torch, 'no_grad'),
  'torch.inference_mode': hasattr(torch, 'inference_mode'),
  'torch.save': hasattr(torch, 'save'),
  'torch.load': hasattr(torch, 'load'),
  'torch.sigmoid': hasattr(torch, 'sigmoid'),
}
try:
    t = torch.tensor([1.0], requires_grad=True)
    caps['Tensor.clone'] = hasattr(t, 'clone')
    caps['Tensor.detach'] = hasattr(t, 'detach')
    caps['Tensor.is_leaf'] = hasattr(t, 'is_leaf')
    caps['Tensor.grad_fn'] = hasattr(t, 'grad_fn')
except Exception as e:
    caps['tensor_probe_error'] = str(e)
json.dumps(caps, sort_keys=True)
`);
console.log(result);
```

Output:

```json
{
  "Tensor.clone": false,
  "Tensor.detach": false,
  "Tensor.grad_fn": false,
  "Tensor.is_leaf": false,
  "nn.BatchNorm1d": false,
  "nn.Dropout": true,
  "torch.inference_mode": false,
  "torch.load": false,
  "torch.no_grad": false,
  "torch.save": false,
  "torch.sigmoid": false,
  "version_context": "@unlocalhosted/browsergrad-jit 0.8.1 via Pyodide"
}
```

## Why these matter

These APIs unlock realistic browser exercises without leaving Pyodide:

- `torch.sigmoid`: binary classifier logits -> probabilities -> threshold reports.
- `Tensor.detach`, `Tensor.clone`, `Tensor.is_leaf`, `Tensor.grad_fn`: teach logging boundaries, leaf tensors, graph debugging, and detach-vs-clone intuition.
- `torch.no_grad` / `torch.inference_mode`: validation and inference loops that do not build graphs.
- `nn.BatchNorm1d`: train/eval mode exercises beyond Dropout, especially running-stat behavior.
- `torch.save` / `torch.load`: checkpoint roundtrip and resume exercises.

## Suggested priority

For education use cases, the highest-leverage order seems:

1. `torch.sigmoid`
2. `Tensor.detach` + `Tensor.clone`
3. `torch.no_grad`
4. `Tensor.is_leaf` / `Tensor.grad_fn` read-only metadata
5. `nn.BatchNorm1d`
6. minimal `torch.save` / `torch.load` for state_dict-shaped dicts

Happy to split this issue if smaller API-specific issues are preferred.

**Comment by vijayksingh** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Category:** enhancement
**Summary:** Finish the remaining TensorProxy curriculum compatibility surface: honest contiguity queries and tensor indexing.

**Current behavior:**
The original sigmoid, autograd metadata, grad-mode contexts, BatchNorm1d, save/load, clone/detach, and CPU device requests are now implemented. A fresh Pyodide check confirms TensorProxy.device reports `cpu`, but TensorProxy has no `is_contiguous()` and is not subscriptable. The reported paired-index example therefore raises `TypeError` instead of producing `[12, 21]`.

**Desired behavior:**
TensorProxy should expose a boolean contiguity query that reflects its represented layout, and indexing should support the paired integer-tensor gather used by browser-first training material. Indexing must remain differentiable: backward should scatter-add into the source shape, including repeated indices. Unsupported indexing forms must fail at the semantic boundary with a specific diagnostic rather than silently realizing or returning different semantics.

**Key interfaces:**
- `TensorProxy.is_contiguous()` — report layout truth without forcing an unrelated device transfer.
- `TensorProxy.__getitem__()` — represent supported basic and paired advanced indexing in the lazy graph.
- Indexing backward — scatter upstream gradients to the selected source positions and preserve dtype/session ownership.

**Acceptance criteria:**
- [ ] `x.is_contiguous()` returns a Python bool and is correct for every currently representable dense layout.
- [ ] For `x=[[10,11,12],[20,21,22]]`, row indices `[0,1]`, and column indices `[2,1]`, `x[rows, cols]` realizes to `[12,21]`.
- [ ] Gradients through supported indexing have the source shape and scatter-add correctly for duplicate indices.
- [ ] Unsupported index dtype, rank, or form produces a clear compatibility error.
- [ ] Existing device, lazy execution, dtype, and torch-alias behavior remains covered and passing.

**Out of scope:**
- Full NumPy/PyTorch indexing parity beyond the forms explicitly represented by the compatibility contract.
- Non-CPU device support or a second source-shaped execution path.

**Comment by vijayksingh** [comment]
> *This was generated by AI during triage.*

## Verified outcome

**State:** wontfix — already implemented

The reported browser-training API surface is implemented: sigmoid, clone/detach and graph metadata, no-grad/inference contexts, BatchNorm1d, and save/load are covered by the torch-alias integration contract. The remaining indexing/contiguity teaching gaps were also completed in https://github.com/unlocalhosted/browsergrad/commit/4d7181a9e14ba8c69e8714277da4172b361afcf4. Fresh verification passed JIT build, both typechecks, lint, 27 unit tests, 345 integration tests, architecture checks, packed-release tests, and the required JIT semantic-permute WebGPU gate.

Closing as completed and replacing the ready state with `wontfix` per the repository triage state machine.
